Rewrite BC editor to be able to add GitHub and Generic webhook triggers

jhadvig opened the pull request:
So the reported issue contained two possible solutions, but I didn't liked the idea of diplaying the webhook URL, since user can not modify the URL anyway, just disable/enable the triggers, so there is no point of  showing the URL.
I've added the `Learn more` link and also the link for the configuration tab, although not sure if we want a redirection link besides the breadcrumbs. On the other hand it could come really handy if user would want to open the `Configuration` tab in a new tab.

Here are two drafts:
1#
![screenshot-61](https://cloud.githubusercontent.com/assets/1668218/17435217/8caa34c4-5b0f-11e6-9ad5-9ed2193d530d.png)

2#
![screenshot-62](https://cloud.githubusercontent.com/assets/1668218/17435218/8faa7800-5b0f-11e6-9e68-1ac1c358f956.png)

Personally I like the first one since the `Learn more` link is kinda weird next to the header.

@jwforres PTAL

Closes https://github.com/openshift/origin/issues/10098

spadgett commented:
If I click the configuration tab link, will I lose any edits I've made? Not sure we should encourage the user to navigate away while editing.

I wouldn't put the (i) icon in the help message since we don't in other places. I'd also just use a plain "Lean more" link under the help text to be consistent with other doc links. I'm thinking of something like this:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/17436871/27443afc-5ae8-11e6-9e2b-3027450f8ee3.png)

> So the reported issue contained two possible solutions, but I didn't liked the idea of diplaying the webhook URL, since user can not modify the URL anyway, just disable/enable the triggers, so there is no point of showing the URL.

Wondering if we shouldn't have a show webhook URL link somewhere to reveal it if I want to see. As I'm editing the build config, I might want to update the GitHub repo at the same time.

jhadvig commented:
![screenshot-64](https://cloud.githubusercontent.com/assets/1668218/17438221/11266bb6-5b22-11e6-80a8-9221e63dcb49.png)

jhadvig commented:
> Wondering if we shouldn't have a show webhook URL link somewhere to reveal it if I want to see. As I'm editing the build config, I might want to update the GitHub repo at the same time.

What about creating having the link to the configuration tab, but force to open it in a new tab, if clicked ?

spadgett commented:
But if I just enabled the webhook, it won't be there in the new tab.

jhadvig commented:
True that. OK lets drop that idea. 

> Wondering if we shouldn't have a show webhook URL link somewhere to reveal it if I want to see. As I'm editing the build config, I might want to update the GitHub repo at the same time.

@spadgett you mean to only show the full URL of the webhook ?

spadgett commented:
I'm thinking of a "Show webhook URLs"  link that will show the copy to clipboard widget with the values directly in the editor.

jhadvig commented:
@spadgett so here is a preview of the updated BC editor.
![screenshot-1](https://cloud.githubusercontent.com/assets/1668218/17584313/18cd4e5a-5fb6-11e6-89e8-b8a2a7d03b1e.png)

I know we talked to put close icon after the copy-to-clipboard widget but I've realized that if user will by any chance accidentally click on the `X` btn then we would get rid of the webhook and the only way how to undo that would be to Cancel from the editor. So instead I've put there `Remove` btn, which will change to `Undo` and will disable the widget, so if user accidentally remove the webhook, or just changes his mind he caneasily undo the change.
Thoughts ?

spadgett commented:
"Remove" looks good. It might have been weird with different button styles for copy and remove anyway if we went with the icon.

I wouldn't disable the copy to clipboard input for new values, and I'd always use "Remove" rather than "Undo".

jwforres commented:
I assumed the disabled with the Undo was after you clicked Remove, is it
not?

On Thu, Aug 11, 2016 at 8:42 AM, Sam Padgett <notifications@github.com>
wrote:

> "Remove" looks good. It might have been weird with different button styles
> for copy and remove anyway if we went with the icon.
>
> I wouldn't disable the copy to clipboard input for new values, and I'd
> always use "Remove" rather than "Undo".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/298#issuecomment-239149620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7RYq3O0e5U_q4DomRzAoRLcQuIXLks5qexi3gaJpZM4Jdl_T>
> .
>

spadgett commented:
> I assumed the disabled with the Undo was after you clicked Remove

Oh must be, I misunderstood.

jhadvig commented:
@jwforres @spadgett correct after user hits the `Remove` btn the copy-to-clipboard widget will get disabled and `Remove` will change to `Undo`
Only the webhooks that wont be disabled will be save.

jhadvig commented:
@spadgett @jwforres I've updated the PR with the logic described in previous comments.
It's tested and no data loss shall occur. 
Although there are two things I would like to align on with you:
1. position of the `Learn more` link. Here are screens of link different positions:
**Right to the section header**
![screenshot-2](https://cloud.githubusercontent.com/assets/1668218/17618652/38fff8be-6081-11e6-83e6-125007bd547f.png)
 **Under to the section header**
![screenshot-3](https://cloud.githubusercontent.com/assets/1668218/17618655/3bcfb4e4-6081-11e6-8c50-1715443aaaa8.png)
I know we discussed this with Sam to place it under the header, but I kinda like the one on the right from the header, since for the headers in different sections we have a tooltip help right to the header.

2. wasn't sure if we want to disable the copy-to-clipboard mechanism the widget is disabled, since user could copy the URL from the input field anyway, even if the widget is disabled. Then the disabling would signal that the webhook will be removed.

PTAL

jwforres commented:
i prefer to learn more to the right of the header personally.  disabling the copy link would be good if the changes to do it arent big

jwforres commented:
If the plan is for this PR to fix https://bugzilla.redhat.com/show_bug.cgi?id=1365650  then make sure you update the create from image flow, we still want a checkbox there to turn on the config change trigger, but it should say something like "Launch a build as soon as the build configuration is created."

jwforres commented:
or "Launch the first build when the build configuration is created."

jhadvig commented:
> If the plan is for this PR to fix https://bugzilla.redhat.com/show_bug.cgi?id=1365650 then make sure you update the create from image flow, we still want a checkbox there to turn on the config change trigger, but it should say something like "Launch a build as soon as the build configuration is created."

@jwforres not really sure what you mean by adding a checkbox that will add configChange trigger to BC, since in the  'create from image flow' a.k.a 'Deploy Image' tab, no BC will be created, only DC, IS and Service.

jwforres commented:
Not the Deploy Image tab, the Browse Catalog tab when you select a builder
image.  The "from source" flow

On Aug 15, 2016 6:25 AM, "Jakub Hadvig" <notifications@github.com> wrote:

> If the plan is for this PR to fix https://bugzilla.redhat.com/
> show_bug.cgi?id=1365650 then make sure you update the create from image
> flow, we still want a checkbox there to turn on the config change trigger,
> but it should say something like "Launch a build as soon as the build
> configuration is created."
>
> @jwforres <https://github.com/jwforres> not really sure what you mean by
> adding a checkbox that will add configChange trigger to BC, since in the
> 'create from image flow' a.k.a 'Deploy Image' tab, no BC will be created,
> only DC, IS and Service.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/298#issuecomment-239772438>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7Y75CwuJ2HwsKwKdzSk3yZGv4ULtks5qgD6vgaJpZM4Jdl_T>
> .
>

jwforres commented:
There is already a checkbox in the create from source flow that has the
wrong text on it based on what we know about the config change trigger now

On Aug 15, 2016 7:37 AM, "Jessica Forrester" <jforrest@redhat.com> wrote:

> Not the Deploy Image tab, the Browse Catalog tab when you select a builder
> image.  The "from source" flow
>
> On Aug 15, 2016 6:25 AM, "Jakub Hadvig" <notifications@github.com> wrote:
>
>> If the plan is for this PR to fix https://bugzilla.redhat.com/sh
>> ow_bug.cgi?id=1365650 then make sure you update the create from image
>> flow, we still want a checkbox there to turn on the config change trigger,
>> but it should say something like "Launch a build as soon as the build
>> configuration is created."
>>
>> @jwforres <https://github.com/jwforres> not really sure what you mean by
>> adding a checkbox that will add configChange trigger to BC, since in the
>> 'create from image flow' a.k.a 'Deploy Image' tab, no BC will be created,
>> only DC, IS and Service.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openshift/origin-web-console/pull/298#issuecomment-239772438>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABZk7Y75CwuJ2HwsKwKdzSk3yZGv4ULtks5qgD6vgaJpZM4Jdl_T>
>> .
>>
>

jhadvig commented:
@jwforres ohh sorry my mistake. Will update as well. Thanks !

jhadvig commented:
@jwforres I've updated the PR and addressed the comments.
I've put the webhook edit login into a directive to reduce the size of _edit/build-config.html_ and _edit/buildConfig.js_ at least a bit.
Also for the mobile version I've added icons describing the action instead of the *Remove* and *Undo* word action, cause on the small screen the word actions would end up on next line. Screens
![screenshot-6](https://cloud.githubusercontent.com/assets/1668218/17665931/1431849c-62fe-11e6-99ae-d4afc107eb07.png)
![screenshot-7](https://cloud.githubusercontent.com/assets/1668218/17665933/16913c32-62fe-11e6-8fb0-1428d38037f9.png)


> i prefer to learn more to the right of the header personally. disabling the copy link would be good if the changes to do it arent big

unfortunately the changes would be need to be bigger, and I would prefer not to do them ATM

 PTAL

jhadvig commented:
Both... but thats just for screens with width 421px and smaller

spadgett commented:
@jhadvig You will have to click the buttons twice on iOS if we leave the tooltip. It's not a good experience. Can you use a `title` attribute instead?

jhadvig commented:
@spadgett yeah I would rather get rid of the tooltip as well, just didn't know how to replace it. Thanks for  hint !

spadgett commented on app/scripts/constants.js:
```diff
@@ -12,6 +12,7 @@ window.OPENSHIFT_CONSTANTS = {
     "cli":                     "https://docs.openshift.org/latest/cli_reference/overview.html",
     "get_started_cli":         "https://docs.openshift.org/latest/cli_reference/get_started_cli.html",
     "basic_cli_operations":    "https://docs.openshift.org/latest/cli_reference/basic_cli_operations.html",
+    "build-triggers":          "https://docs.openshift.org/latest/dev_guide/builds.html#build-triggers",
```
@jwforres @sg00dwin FYI, we'll need to update the enterprise-3.3 branch as well for the new doc link

spadgett commented on app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -335,42 +333,56 @@ angular.module('openshiftConsole')
+          present: isPresent,
+          data: trigger
+        };
+      }
```
I think these functions are simple enough I'd just inline them.

spadgett commented on app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -671,49 +683,16 @@ angular.module('openshiftConsole')
+                              $scope.triggers.configChangeTrigger);
+      // Filter webhook triggers that are not disabled or imageChange triggers that are present
+      triggers = _.filter(triggers, function(trigger) {
+        return trigger.disabled === false || trigger.present;
```
```js
return !trigger.disabled || trigger.present;
```

jhadvig replied:
We need to check if the `trigger.disabled` equals `false`, cause if the trigger wont have `disabled` field set  (eg. imageChange trigger) it will evaluate as `true`, which is not correct.

spadgett replied:
Oh, I see. I think we need to make it more clear in that case. Something like

```js
if (_.has(trigger, 'disabled') && !trigger.disabled) {
  return true;
}
return trigger.present;
```

Then it's more obvious that you're only checking this value if it exists.

It might be better even to have a switch on the different types and only check the property (disabled or present) that we expect.

spadgett replied:
Can you add a comment explaining the `trigger.disabled === false` check?

spadgett commented on app/scripts/directives/editWebhookTriggers.js:
```diff
@@ -0,0 +1,33 @@
+              type: type
+            }
+          };
+          webhook.data[type.toLowerCase()] = {
```
I think I'd rather check for known types instead of assuming that `toLowerCase()` will always give the right property name.

spadgett commented on app/scripts/directives/util.js:
```diff
@@ -75,19 +75,21 @@ angular.module('openshiftConsole')
       templateUrl: 'views/directives/_copy-to-clipboard.html',
       controller: function($scope) {
         $scope.id = _.uniqueId('clipboardJs');
+        $scope.isDisabled = ($scope.isDisabled) ? true : false;
```
Is this needed?

spadgett commented on app/styles/_core.less:
```diff
@@ -1091,6 +1098,23 @@ a.disabled-link {
+      margin-left: 5px;
+      .action-icon,
+      .action-icon:hover {
+        color: #4d5258;
```
```css
color: @color-pf-black-700;
```

spadgett commented on app/views/directives/_copy-to-clipboard.html:
```diff
@@ -1,9 +1,10 @@
   <span class="input-group-btn" ng-hide="hidden">
-    <button data-clipboard-target="#{{id}}"
+    <a data-clipboard-target="#{{id}}"
+            ng-disabled="isDisabled"
```
Does `ng-disabled` work on links? I didn't think it did.

jhadvig replied:
no it wont disable the link. It's just to graphically distinguish the widget in disabled and not-disabled state. we already spoke about it with @jwforres https://github.com/openshift/origin-web-console/pull/298#issuecomment-239510865

spadgett commented on app/views/directives/_copy-to-clipboard.html:
```diff
@@ -1,9 +1,10 @@
             data-toggle="tooltip"
             title="Copy to clipboard"
-            class="btn btn-default"><i class="fa fa-clipboard"/></button>
+            class="btn btn-default"><i class="fa fa-clipboard"/></a>
```
I'd also add `role="button"` for accessibility

spadgett commented on app/views/directives/edit-webhook-triggers.html:
```diff
@@ -0,0 +1,30 @@
+<h5>{{type}} webhooks
+  <span class="help action-inline">
+    <a href>
+      <i class="pficon pficon-help" data-toggle="tooltip" aria-hidden="true" data-original-title="{{typeInfo}}">
```
I'd also add the text for screen readers in an sr-only span.

```html
<span class="sr-only">{{typeInfo}}</span>
```

spadgett replied:
I'd also put the aria-hidden on the parent `a` element. Otherwise screen readers will read it as an empty link.

spadgett commented on app/views/directives/edit-webhook-triggers.html:
```diff
@@ -0,0 +1,30 @@
+    </span>
+    <span class="visible-xs-inline trigger-actions">
+      <a href="" ng-if="!trigger.disabled" class="action-icon" ng-click="trigger.disabled = true; form.$setDirty()">
+        <span class="pficon pficon-close" aria-hidden="true" data-toggle="tooltip" data-original-title="Remove"></span>
```
Add an sr-only span with the text for screenreaders

spadgett commented on app/views/directives/edit-webhook-triggers.html:
```diff
@@ -0,0 +1,30 @@
+        <span class="pficon pficon-close" aria-hidden="true" data-toggle="tooltip" data-original-title="Remove"></span>
+      </a>
+      <a href="" ng-if="trigger.disabled" class="action-icon" ng-click="trigger.disabled = false">
+        <span class="fa fa-repeat" aria-hidden="true" data-toggle="tooltip" data-original-title="Undo"></span>
```
Needs an sr-only span here too.

spadgett commented on app/views/directives/edit-webhook-triggers.html:
```diff
@@ -0,0 +1,30 @@
+      <copy-to-clipboard is-disabled="trigger.disabled" clipboard-text="bcName | webhookURL : trigger.data.type : trigger.data[type.toLowerCase()].secret : projectName"></copy-to-clipboard>
+    </span>
+    <span class="visible-xs-inline trigger-actions">
+      <a href="" ng-if="!trigger.disabled" class="action-icon" ng-click="trigger.disabled = true; form.$setDirty()">
```
role="button"

spadgett commented on app/views/directives/edit-webhook-triggers.html:
```diff
@@ -0,0 +1,30 @@
+      <a href="" ng-if="!trigger.disabled" class="action-icon" ng-click="trigger.disabled = true; form.$setDirty()">
+        <span class="pficon pficon-close" aria-hidden="true" data-toggle="tooltip" data-original-title="Remove"></span>
+      </a>
+      <a href="" ng-if="trigger.disabled" class="action-icon" ng-click="trigger.disabled = false">
```
role="button"

spadgett commented on app/views/directives/edit-webhook-triggers.html:
```diff
@@ -0,0 +1,30 @@
+    </span>
+    <span class="hidden-xs trigger-actions">
+      <a href="" ng-if="!trigger.disabled" ng-click="trigger.disabled = true; form.$setDirty()">Remove</a>
+      <a href="" ng-if="trigger.disabled" ng-click="trigger.disabled = false">Undo</a>
```
role="button" on these link as well

spadgett commented on app/views/directives/edit-webhook-triggers.html:
```diff
@@ -0,0 +1,30 @@
+  </div>
+</div>
+<span class="learn-more-inline checkbox">
+  <a href="" ng-click="addWebhookTrigger(type)">Add GitHub webhook</a>
```
role="button"

spadgett replied:
Shouldn't this be `Add {{type}} webhook`?

spadgett commented on app/views/edit/build-config.html:
```diff
@@ -518,56 +518,55 @@
                       <div class="section">
-                        <h3>Triggers</h3>
+                        <h3>Triggers
+                          <a href="{{'build-triggers' | helpLink}}" target="_blank"><span class="learn-more-inline">Learn more<i class="fa fa-external-link"></i></span></a>
```
`aria-hidden="true"` on the icon

spadgett commented:
@jhadvig code looks much cleaner than before moving things into a directive. thanks

spadgett commented on app/views/directives/edit-webhook-triggers.html:
```diff
@@ -0,0 +1,30 @@
+<div ng-repeat="trigger in triggers">
+  <div class="trigger-info">
+    <span class="trigger-url">
+      <copy-to-clipboard is-disabled="trigger.disabled" clipboard-text="bcName | webhookURL : trigger.data.type : trigger.data[type.toLowerCase()].secret : projectName"></copy-to-clipboard>
```
Same about about `toLowerCase` as I made in the controller.

jhadvig commented:
@spadgett comments addressed. PTAL

spadgett commented on app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -335,42 +333,50 @@ angular.module('openshiftConsole')
+      // Applies to image and config change triggers
+      function constructChangeTrigger(trigger, isPresent) {
+        return {present: isPresent, data: trigger};
+      }
```
Sorry, by "inline" I meant not have the functions, just create the objects below.

spadgett commented:
thanks @jhadvig 

[merge]
